moonraker/components/filament_manager.py

```diff
@@ -0,0 +1,296 @@
+from typing import Optional
+from typing import List
+from components.database import NamespaceWrapper
+
+WebRequest = "websockets.WebRequest"  # type mapping
+
+FILAMENT_NAMESPACE = "filament_manager"
+SPOOLS_SUB_NS = "spools"
+ACTIVE_SPOOL_KEY = "active-spool"
+
+MATERIALS = {
+    'PLA': {'density': 1.24},
+    'PLA_plus': {'density': 1.24},
+    'ABS': {'density': 1.04},
+    'PETG': {'density': 1.27},
+    'NYLON': {'density': 1.52},
+    'TPU': {'density': 1.21},
+    'PC': {'density': 1.3},
+    'Carbon': {'density': 1.3},
+    'PC_ABS': {'density': 1.19},
+    'HIPS': {'density': 1.03},
+    'PVA': {'density': 1.23},
+    'ASA': {'density': 1.05},
+    'PP': {'density': 0.9},
+    'POM': {'density': 1.4},
+    'PMMA': {'density': 1.18},
+    'FPE': {'density': 2.16}
+}
+
+
+class Validation:
+    def validate(self) -> Set[str]:
+        failed = filter(lambda f: self.__getattribute__(f) is None,
+                        self._required_attributes)
+        return set(failed)
+
+
+class Spool(Validation):
+    _required_attributes: Set[str] = {'name'}
+
+    def __init__(self, data={}):
+        self.name: str = None
+        self.active: bool = True
+        self.color_name: str = None
+        self.color_code: str = None
+        self.vendor: str = None
+        self.material: str = None
+        self.density: float = None
+        self.diameter: float = None
+        self.total_weight: int = None
+        self.used_weight: int = 0
+        self.total_length: int = None
+        self.used_length: int = 0
+        self.first_used: float = None
+        self.last_used: float = None
+        self.cost: float = None
+        self.comment: str = None
+
+        self.update(data)
+
+    def update(self, data):
+        for a in data:
+            if hasattr(self, a):
+                setattr(self, a, data[a])
+
+    def serialize(self):
+        return self.__dict__.copy()
+
+
+class FilamentManager:
+    def __init__(self, config):
+        self.server = config.get_server()
+
+        database = self.server.lookup_component("database")
+        database.register_local_namespace(FILAMENT_NAMESPACE)
+        self.db: NamespaceWrapper = database.wrap_namespace(FILAMENT_NAMESPACE,
+                                                            parse_keys=True)
+
+        self.endpoint = FilamentManagerEndpoint(self.server, self)
+        self.listener = FilamentManagerListener(self.server, self)
+
+    def _spool_id(self, spool_id: str) -> str:
+        return f"{SPOOLS_SUB_NS}.{spool_id}"
+
+    def find_spool(self, spool_id: str, default = None) -> Optional[dict]:
+        return self.db.get(self._spool_id(spool_id), default)
+
+    def set_active_spool(self, spool_id: str) -> Optional[dict]:
+        spool = self.find_spool(spool_id)
+
+        if spool:
+            if self.db.get(ACTIVE_SPOOL_KEY):
+                self.db.update_child(ACTIVE_SPOOL_KEY, spool_id)
+            else:
+                self.db.insert(ACTIVE_SPOOL_KEY, spool_id)
+
```

Comment:
updated that,
Yep, wrapped namespace use mapping protocol, that is nice, but unfortunately __getitem__ returns Sentinel as defalut instead of None as the base get() method.

Reply:
Right, if the key doesn't exist an access using the mapping protocol will raise an exception. It is fine to use the `get()` method if you want a default value.